zerver/lib/topic_sqlalchemy.py

```diff
@@ -18,3 +19,22 @@ def get_resolved_topic_condition_sa() -> ColumnElement[Boolean]:
 
 def topic_column_sa() -> ColumnElement[Text]:
     return column("subject", Text)
+
+
+def get_followed_topic_condition_sa(user_id: int) -> ColumnElement[Boolean]:
+    follow_topic_cond = (
+        select([1])
+        .select_from(table("zerver_usertopic"))
+        .where(
+            and_(
+                literal_column("zerver_usertopic.user_profile_id") == literal(user_id),
+                literal_column("zerver_usertopic.visibility_policy")
+                == literal(UserTopic.VisibilityPolicy.FOLLOWED),
+                literal_column("zerver_usertopic.topic_name")
+                == literal_column("zerver_message.subject"),
```

Comment:
This check is wrong -- topics are case-insensitive.

Reply:
`topic_match_sa` in this same file has correct logic you can use.

Can you also post the SQL that this generates, and its query plan from `EXPLAIN ANALYZE`? We may need to adjust some indexes to make this efficient.

Reply:
> This check is wrong -- topics are case-insensitive.

Ok, a related issue that I noticed:
If I try to create a new topic with the same name but different case, duplicate topics show up in the sidebar.
All messages are sent to the original topic itself and if i collapse the sidebar and reopen, the duplicate topic disappears.
![image](https://github.com/zulip/zulip/assets/29176437/4bf09933-d2f4-4361-b3ee-0079debcf2ce)

Reply:
Query
```
SELECT message_id, flags
FROM zerver_usermessage
JOIN zerver_message ON zerver_usermessage.message_id = zerver_message.id
WHERE user_profile_id = 33 
AND EXISTS (
    SELECT 1 
    FROM zerver_usertopic 
    WHERE zerver_usertopic.user_profile_id = 33 
    AND zerver_usertopic.visibility_policy = 3 
AND upper(zerver_usertopic.topic_name) = upper(zerver_message.subject)
    AND zerver_usertopic.recipient_id = zerver_message.recipient_id
) 
ORDER BY message_id DESC
LIMIT 50;

```
EXPLAIN ANALYZE

```
                                                                                             QUERY PLAN

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=265.27..265.28 rows=1 width=12) (actual time=1.442..1.449 rows=50 loops=1)
   ->  Sort  (cost=265.27..265.28 rows=1 width=12) (actual time=1.440..1.443 rows=50 loops=1)
         Sort Key: zerver_usermessage.message_id DESC
         Sort Method: top-N heapsort  Memory: 29kB
         ->  Nested Loop  (cost=256.96..265.26 rows=1 width=12) (actual time=0.311..1.399 rows=170 loops=1)
               ->  Nested Loop  (cost=256.67..264.45 rows=1 width=4) (actual time=0.301..0.964 rows=170 loops=1)
                     ->  HashAggregate  (cost=1.00..1.01 rows=1 width=142) (actual time=0.023..0.026 rows=3 loops=1)
                           Group Key: upper((zerver_usertopic.topic_name)::text), zerver_usertopic.recipient_id
                           Batches: 1  Memory Usage: 24kB
                           ->  Seq Scan on zerver_usertopic  (cost=0.00..1.00 rows=1 width=142) (actual time=0.015..0.017 rows=3 loops=1)
                                 Filter: ((user_profile_id = 30) AND (visibility_policy = 3))
                                 Rows Removed by Filter: 2
                     ->  Bitmap Heap Scan on zerver_message  (cost=255.66..263.42 rows=2 width=27) (actual time=0.200..0.305 rows=57 loops=3)
                           Recheck Cond: ((recipient_id = zerver_usertopic.recipient_id) AND (upper((subject)::text) = upper((zerver_usertopic.topic_name)::text)))
                           Heap Blocks: exact=157
                           ->  BitmapAnd  (cost=255.66..255.66 rows=2 width=0) (actual time=0.184..0.184 rows=0 loops=3)
                                 ->  Bitmap Index Scan on zerver_message_recipient_id_5a7b6f03  (cost=0.00..5.62 rows=179 width=0) (actual time=0.013..0.013 rows=221 loops=3)
                                       Index Cond: (recipient_id = zerver_usertopic.recipient_id)
                                 ->  Bitmap Index Scan on zerver_message_realm_upper_subject  (cost=0.00..249.78 rows=52 width=0) (actual time=0.160..0.160 rows=57 loops=3)
                                       Index Cond: (upper((subject)::text) = upper((zerver_usertopic.topic_name)::text))
               ->  Index Scan using zerver_usermessage_user_profile_id_message_id_4936d0df_uniq on zerver_usermessage  (cost=0.29..0.81 rows=1 width=12) (actual time=0.002..0.002 rows=1 loops=170)
                     Index Cond: ((user_profile_id = 30) AND (message_id = zerver_message.id))
 Planning Time: 0.901 ms
 Execution Time: 1.578 ms
(24 rows)
```

Reply:
One should always cross-link threads in both directions. This was:
https://chat.zulip.org/#narrow/stream/3-backend/topic/is.3Afollowed.20operator/near/1820472

Where we decided the query is probably fine as written.